chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use common::{
+    chain::{timelock::OutputTimeLock, AccountSpending, ChainConfig, PoolId, TxInput, TxOutput},
+    primitives::{Amount, BlockDistance, BlockHeight},
+};
+use utils::ensure;
+
+use super::IOPolicyError;
+
+/// `ConstrainedValueAccumulator` helps avoiding messy inputs/outputs combinations analysis by
+/// providing a set of properties that should be satisfied. For example instead of checking that
+/// all outputs are timelocked when the pool is decommissioned `ConstrainedValueAccumulator` gives a way
+/// to check that an accumulated output value is locked for sufficient amount of time which allows
+/// using other valid inputs and outputs in the same tx.
+///
+/// TODO: this struct can be extended to collect tokens replacing `AmountsMap`
+pub struct ConstrainedValueAccumulator {
+    unconstrained_value: Amount,
+    timelock_constrained: BTreeMap<BlockDistance, Amount>,
```

Comment:
We should probably create a NonZeroBlockDistance for this. But maybe later.

Reply:
I don't think that we should treat BlockDistance(0) as a separate case. It cannot satisfy constrained value so we just subtract the value from unconstrained. The same way we would subtract BlockDistance(1) or any other value less then required by maturity.

Reply:
Later when we apply this to fees it'll become apparent why this is necessary.